File: nodejs/backend/src/controller/authController.ts
import { Request, Response } from 'express';
import { v4 as uuidv4 } from 'uuid';
import { AuthService } from '../service/AuthService';
import { toUserResponseDto, LoginResponseDto } from '../dto/authDto';

export class AuthController {
  constructor(private readonly authService: AuthService) {}

  signup = async (req: Request, res: Response): Promise<void> => {
    // req.body was already parsed + validated by validateBody(signupSchema)
    const user = await this.authService.signup(req.body);
    res.status(201).json(toUserResponseDto(user));
  };

  login = async (req: Request, res: Response): Promise<void> => {
    const user = await this.authService.login(req.body);
    /*
      NOTE: this issues a random UUID as a "token" with no signing, expiry, or session storage. 
      It is not a real bearer token yet. Wire up real auth (JWT or a session table)
      before protecting any route with it.
    */
    const response: LoginResponseDto = {
      token: uuidv4(),
      user: toUserResponseDto(user),
    };

    res.json(response);
  };
}

File: rust/backend/src/controller/auth_controller.rs
use crate::{
    dto::{LoginRequest, LoginResponse, SignupRequest, UserResponse},
    error::AppError,
    state::AppState,
};
use axum::{extract::State, http::StatusCode, response::IntoResponse, Json};
use uuid::Uuid;

pub async fn signup(
    State(state): State<AppState>,
    Json(payload): Json<SignupRequest>,
) -> Result<impl IntoResponse, AppError> {
    let user = state.auth_service.signup(payload).await?;
    Ok((StatusCode::CREATED, Json(UserResponse::from(user))))
}

pub async fn login(
    State(state): State<AppState>,
    Json(payload): Json<LoginRequest>,
) -> Result<impl IntoResponse, AppError> {
    let user = state.auth_service.login(payload).await?;

    // NOTE: this issues a random UUID as a "token" with no signing, expiry,
    // or session storage — same as the original code. It is not a real
    // bearer token. See the note in middleware/auth.rs for what's needed
    // before this is safe to use as an auth mechanism.
    let response = LoginResponse {
        token: Uuid::new_v4().to_string(),
        user: UserResponse::from(user),
    };

    Ok(Json(response))
}
